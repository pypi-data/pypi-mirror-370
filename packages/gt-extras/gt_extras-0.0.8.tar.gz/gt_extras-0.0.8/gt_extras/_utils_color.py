from __future__ import annotations

import warnings

from great_tables._data_color.base import (
    _get_domain_factor,
    _html_color,
    _rescale_factor,
)
from great_tables._data_color.constants import ALL_PALETTES, DEFAULT_PALETTE
from great_tables._data_color.palettes import GradientPalette
from great_tables._tbl_data import TblData

__all__ = ["_get_discrete_colors_from_palette"]


def _get_discrete_colors_from_palette(
    palette: list[str] | str | None,
    data: list,
    data_table: TblData,
) -> list[str]:
    palette = _get_palette(palette)

    # Rescale the category column for the purpose of assigning colors to each dot
    domain_factor = _get_domain_factor(df=data_table, vals=data)
    scaled_vals = _rescale_factor(
        df=data_table,
        vals=data,
        domain=domain_factor,  # type: ignore
        palette=palette,
    )

    # Create a color scale function from the palette
    color_scale_fn = GradientPalette(colors=palette)

    # Call the color scale function on the scaled categoy values to get a list of colors
    color_vals = color_scale_fn(scaled_vals)

    for i, c in enumerate(color_vals):
        if c is None:
            color_vals[i] = "transparent"
            warnings.warn(
                "A color value is None and has been coerced to 'transparent'",
                UserWarning,
            )

    # This is redundant but satisfies the type-checker
    color_vals = [c for c in color_vals if c is not None]

    return color_vals


def _get_palette(palette: list[str] | str | None) -> list[str]:
    # If palette is not provided, use a default palette
    if palette is None:
        palette = DEFAULT_PALETTE

    # Otherwise get the palette from great_tables._data_color
    elif isinstance(palette, str):
        palette = ALL_PALETTES.get(palette, [palette])

    # Standardize values in `palette` to hexadecimal color values
    palette = _html_color(colors=palette)

    return palette
