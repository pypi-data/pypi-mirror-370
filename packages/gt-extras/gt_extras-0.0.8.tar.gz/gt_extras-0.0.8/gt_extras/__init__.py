# Import objects from the module
from .colors import (
    gt_color_box,
    gt_data_color_by_group,
    gt_highlight_cols,
    gt_highlight_rows,
    gt_hulk_col_numeric,
)
from .formatting import fmt_pct_extra, gt_duplicate_column, gt_two_column_layout
from .html import gt_merge_stack, with_hyperlink, with_tooltip
from .icons import fa_icon_repeat, gt_fa_rank_change, gt_fa_rating
from .images import add_text_img, gt_fmt_img_circle, img_header
from .plotting import (
    gt_plt_bar,
    gt_plt_bar_pct,
    gt_plt_bar_stack,
    gt_plt_bullet,
    gt_plt_conf_int,
    gt_plt_donut,
    gt_plt_dot,
    gt_plt_dumbbell,
    gt_plt_winloss,
)
from .styling import gt_add_divider
from .summary import gt_plt_summary
from .themes import (
    gt_theme_538,
    gt_theme_dark,
    gt_theme_dot_matrix,
    gt_theme_espn,
    gt_theme_excel,
    gt_theme_guardian,
    gt_theme_nytimes,
    gt_theme_pff,
)

__all__ = [
    "gt_theme_538",
    "gt_theme_espn",
    "gt_theme_nytimes",
    "gt_theme_guardian",
    "gt_theme_excel",
    "gt_theme_dot_matrix",
    "gt_theme_dark",
    "gt_theme_pff",
    "gt_data_color_by_group",
    "gt_highlight_cols",
    "gt_highlight_rows",
    "gt_hulk_col_numeric",
    "gt_color_box",
    "fa_icon_repeat",
    "gt_fa_rank_change",
    "gt_fa_rating",
    "gt_plt_bar",
    "gt_plt_bar_pct",
    "gt_plt_bullet",
    "gt_plt_dot",
    "gt_plt_conf_int",
    "gt_plt_dumbbell",
    "gt_plt_donut",
    "gt_plt_winloss",
    "gt_plt_bar_stack",
    "with_hyperlink",
    "with_tooltip",
    "gt_merge_stack",
    "fmt_pct_extra",
    "gt_duplicate_column",
    "gt_two_column_layout",
    "add_text_img",
    "img_header",
    "gt_fmt_img_circle",
    "gt_add_divider",
    "gt_plt_summary",
]
