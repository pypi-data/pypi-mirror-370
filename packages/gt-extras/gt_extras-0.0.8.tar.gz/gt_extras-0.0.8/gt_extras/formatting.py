from __future__ import annotations

import re
import tempfile
import webbrowser
from pathlib import Path
from typing import TYPE_CHECKING, Literal

import narwhals.stable.v1 as nw
from great_tables import GT
from great_tables._export import _create_temp_file_server
from great_tables._gt_data import Boxhead, ColInfo
from great_tables._tbl_data import SelectExpr, is_na

from gt_extras._utils_column import _validate_and_get_single_column

__all__ = ["fmt_pct_extra", "gt_duplicate_column", "gt_two_column_layout"]


def fmt_pct_extra(
    gt: GT,
    columns: SelectExpr,
    scale: float = 100,
    threshold: float = 1.0,
    color: str = "grey",
    decimals: int = 1,
) -> GT:
    # TODO: would we want to infer the range? (as we do in gt_plt_bar_pct)
    """
    Convert to percent and show less than `1%` as `<1%` in grey.

    The `fmt_pct_extra()` function takes an existing `GT` object and formats a column of numeric
    values as percentages. Values below the specified threshold are displayed as `"<threshold%"`
    instead of their actual percentage value, and in a color of your choosing. The default behavior
    expects values in the range `[0, 1]`, use `scale` if your values are in a different range.

    Parameters
    ----------
    gt
        A `GT` object to modify.

    columns
        The columns containing numeric values to format as percentages.

    scale
        Multiplication factor to convert values to percentages. For some common cases:
        use `100` if values are decimals `(0.05 -> 5%)` (default),
        use `1` if values are already percentages `(5 -> 5%)`.

    threshold
        The percentage threshold below which values are displayed as `"<threshold%"` instead of
        their actual value. Note this refers to the scaled value, not the original, so you will
        have to provide `5` for `<5%` even if your original are in the range `[0,1]`.

    color
        The color to use for values below the threshold.

    decimals
        Number of decimal places to display for percentages.

    Returns
    -------
    GT
        A `GT` object with formatted percentage column.

    Examples
    --------
    ```{python}
    from great_tables import GT
    from great_tables.data import towny
    import gt_extras as gte

    towny_mini = towny[
        [
            "name",
            "pop_change_1996_2001_pct",
            "pop_change_2001_2006_pct",
            "pop_change_2006_2011_pct",
        ]
    ].tail(10)

    gt = (
        GT(towny_mini)
        .tab_spanner(label="Population Change", columns=[1, 2, 3])
        .cols_label(
            pop_change_1996_2001_pct="'96-'01",
            pop_change_2001_2006_pct="'01-'06",
            pop_change_2006_2011_pct="'06-'11",
        )
    )

    gt.pipe(
        gte.fmt_pct_extra,
        columns=[1, 2, 3],
        threshold=5,
        color="green",
    )
    ```
    """
    # TODO: consider how to handle negative values

    def _fmt_pct_single_val(value: float):
        if is_na(gt._tbl_data, value):
            return ""

        # Convert to percentage
        pct_value = value * scale

        if abs(pct_value) < threshold:
            return f"<span style='color:{color};'><{threshold:g}%</span>"
        else:
            return f"{pct_value:.{decimals}f}%"

    res = gt
    res = res.fmt(_fmt_pct_single_val, columns=columns)

    return res


def gt_duplicate_column(
    gt: GT,
    column: SelectExpr,
    after: str | None = None,
    append_text: str | None = "_dupe",
    dupe_name: str | None = None,
) -> GT:
    """
    Duplicate a column in a `GT` object.

    The `gt_duplicate_column()` function takes an existing `GT` object and creates a duplicate
    _(without styling)_ of the specified column. The duplicated column can be renamed using either
    `dupe_name` or by appending text to the original column name, and positioned at a
    specific location in the table.

    Parameters
    ----------
    gt
        A `GT` object to modify.

    column
        The column to duplicate. Can be a column name or index.

    after
        The column after which to place the duplicated column. If `None`, the duplicated
        column will be moved to the end of the table.

    append_text
        Text to append to the original column name for the duplicate. Only used if
        `dupe_name` is not provided. Defaults to `"_dupe"`.

    dupe_name
        The name for the duplicated column. If provided, this overrides `append_text`.

    Returns
    -------
    GT
        A `GT` object with the duplicated column added.

    Examples
    --------
    ```{python}
    from great_tables import GT
    from great_tables.data import gtcars
    import gt_extras as gte

    gtcars_mini = gtcars[["mfr", "model", "year", "hp"]].head(5)
    gt = GT(gtcars_mini)

    # Duplicate with custom name and position
    gt.pipe(
        gte.gt_duplicate_column,
        column="hp",
        after="year",
    )
    ```

    Note
    --------
    Styles generated by [`GT.tab_style()`](https://posit-dev.github.io/great-tables/reference/GT.tab_style)
    will render in the duplicated column. This may be accommodated in the future.
    """
    original_name, _ = _validate_and_get_single_column(gt, column)

    # If dupe_name is given, it overrides append_text
    append_text = append_text or "_dupe"
    if dupe_name is not None:
        new_col_name = dupe_name
    else:
        new_col_name = original_name + append_text

    if new_col_name == original_name:
        raise ValueError(
            f"The new column name '{new_col_name}' cannot be the same as the original column name '{original_name}'."
        )

    res = gt

    # get the boxhead info
    original_col_info = None
    for col_info in res._boxhead:
        if col_info.var == original_name:
            original_col_info = col_info
            break

    if TYPE_CHECKING:
        if original_col_info is None:
            raise ValueError(f"Column '{original_name}' not found in boxhead.")

    # make the new boxhead entry
    new_col_info = ColInfo(
        var=new_col_name,
        type=original_col_info.type,
        column_label=new_col_name,
        column_align=original_col_info.column_align,
        column_width=original_col_info.column_width,
    )

    # Duplicate the col in the data_table data
    original_data_table = res._tbl_data
    new_data_table = nw.to_native(
        nw.from_native(original_data_table).with_columns(
            nw.col(original_name).alias(new_col_name)
        )
    )

    # Duplicate the col in the body element of body
    new_body = res._body
    new_body.body = nw.to_native(
        nw.from_native(new_body.body).with_columns(
            nw.col(original_name).alias(new_col_name)
        )
    )

    new_boxhead_list = list(res._boxhead._d) + [new_col_info]
    new_boxhead = Boxhead(new_boxhead_list)

    res = res._replace(_tbl_data=new_data_table, _boxhead=new_boxhead, _body=new_body)

    if after is None:
        res = res.cols_move_to_end(new_col_name)
    else:
        res = res.cols_move(new_col_name, after=after)

    return res


class GTCombinedLayout:
    """
    Represents the combined layout of two `GT` objects rendered side by side.

    This class is returned by [`gt_two_column_layout()`](https://posit-dev.github.io/gt-extras/reference/gt_two_column_layout)
    and is designed for automatic rendering in Jupyter notebooks and other HTML-aware environments.

    Parameters
    ----------
    html_content
        The combined HTML content representing the two tables in a single layout.

    Methods
    -------
    _repr_html_()
        Returns the HTML content for notebook rendering.

    __str__()
        Returns the HTML content as a string.

    show()
        Displays the table in an environment of choice. Options are `"notebook"` and `"browser"`.
    """

    def __init__(self, html_content: str):
        self.html_content = html_content

    def _repr_html_(self):
        return self.html_content

    def __str__(self):
        return self.html_content

    # do a show based on code from great_tables `GT.show()`
    def show(self, target: Literal["notebook", "browser"] = "browser"):
        if target == "notebook":
            from IPython.core.display import display_html

            # https://github.com/ipython/ipython/pull/10962
            display_html(  # pyright: ignore[reportUnknownVariableType]
                self._repr_html_(), raw=True, metadata={"text/html": {"isolated": True}}
            )
        elif target == "browser":
            with tempfile.TemporaryDirectory() as tmp_dir:
                f_path = Path(tmp_dir) / "index.html"
                f_path.write_text(self._repr_html_(), encoding="utf-8")

                # create a server that closes after 1 request ----
                server = _create_temp_file_server(f_path)
                webbrowser.open(f"http://127.0.0.1:{server.server_port}/{f_path.name}")
                server.handle_request()
        else:
            raise Exception(f"Unknown target display: {target}")

    def save(self):
        raise NotImplementedError


def gt_two_column_layout(
    gt1: GT,
    gt2: GT,
    target: Literal["notebook", "browser"] | None = None,  # TODO: add save
    table_header_from: Literal[1, 2] | None = None,
) -> GTCombinedLayout:
    """
    Combine two `GT` objects into a two-column layout.

    This function takes two `GT` objects and arranges them side by side in a single HTML output.
    Optionally, the header (title and subtitle) from one of the tables can be included at the top
    of the combined layout. The output can be displayed directly in a Jupyter notebook, opened in
    a browser, or returned as HTML.

    Parameters
    ----------
    gt1
        The first table to include in the layout. This table will appear on the left side.

    gt2
        The second table to include in the layout. This table will appear on the right side.

    target
        Determines how the combined layout is displayed. Use `"notebook"` to display in a Jupyter
        notebook, `"browser"` to open in a web browser. If `None`, returns the layout object
        for automatic rendering. Currently, `"save"` is not yet implemented.

    table_header_from
        Determines which table's header (title and subtitle) to include in the combined layout.
        If set to `1`, the header from `gt1` will be used. If set to `2`, the header from `gt2`
        will be used. If `None`, no header will be included.

    Returns
    -------
    GTCombinedLayout
        A layout object that automatically renders as HTML in Jupyter notebooks and Quarto
        documents. It also has a `.show()` method, taking `"notebook"` or `"browser"` as targets.

    Examples
    --------
    ```{python}
    from great_tables import GT
    import gt_extras as gte
    import pandas as pd

    df1 = pd.DataFrame({"A": [1, 2, 3], "B": [4, 5, 6]})
    df2 = pd.DataFrame({"X": [7, 8, 9], "Y": [10, 11, 12]})

    gt1 = GT(df1).tab_header(title="Table 1", subtitle="1st Table")
    gt2 = GT(df2).tab_header(title="Table 2", subtitle="2nd Table")

    gte.gt_two_column_layout(gt1, gt2)
    ```

    Alternatively, we can combine data from the same table in a two column layout like so:

    ```{python}
    df1 = pd.DataFrame({"A": [1, 2, 3], "B": [4, 5, 6]})
    df2 = pd.DataFrame({"A": [7, 8, 9], "B": [10, 11, 12]})

    gt1 = GT(df1).tab_header(title="Joined Table")
    gt2 = GT(df2)

    gte.gt_two_column_layout(gt1, gt2, table_header_from=1)
    ```


    """

    def extract_tab_header_and_style(gt: GT) -> dict:
        """
        Extract the title, subtitle, and full style block from a GT object's HTML.
        """
        html = gt.as_raw_html()

        title = gt._heading.title if gt._heading.title else ""
        subtitle = gt._heading.subtitle if gt._heading.subtitle else ""

        # Extract the original table ID
        id_match = re.search(r'<div id="([^"]*)"', html)
        original_id = id_match.group(1) if id_match else ""

        # Extract the entire style block
        style_match = re.search(r"<style>(.*?)</style>", html, re.DOTALL)
        style_content = style_match.group(1) if style_match else ""

        # Replace the original ID with "mycombinedtable" in the CSS
        if original_id:
            style_content = style_content.replace(f"#{original_id}", "#mycombinedtable")
            style_content = style_content.replace(
                "mycombinedtable table", "mycombinedtable div"
            )

        title_td_pattern = (
            r'<td[^>]*class="[^"]*gt_title[^"]*"[^>]*>\s*'
            + re.escape(str(title))
            + r"\s*</td>"
        )
        title_match = re.search(title_td_pattern, html, re.DOTALL)

        if title_match:
            td_tag = title_match.group(0)

            # Extract class attribute
            class_match = re.search(r'class="([^"]*gt_title[^"]*)"', td_tag)
            title_class = f"gt_table {class_match.group(1)}" if class_match else ""

            # Extract style attribute
            style_match = re.search(r'style="([^"]*)"', td_tag)
            title_inline_style = style_match.group(1) if style_match else ""
        else:
            title_class = ""
            title_inline_style = ""

        subtitle_class = ""

        subtitle_td_pattern = (
            r'<td[^>]*class="[^"]*gt_subtitle[^"]*"[^>]*>\s*'
            + re.escape(str(subtitle))
            + r"\s*</td>"
        )
        subtitle_match = re.search(subtitle_td_pattern, html, re.DOTALL)

        if subtitle_match:
            td_tag = subtitle_match.group(0)

            # Extract class attribute
            class_match = re.search(r'class="([^"]*gt_subtitle[^"]*)"', td_tag)

            if class_match:
                # Exclude gt_bottom_border from class string if present
                class_str = class_match.group(1)
                class_str = " ".join(
                    c for c in class_str.split() if c != "gt_bottom_border"
                )
                subtitle_class = f"gt_table {class_str}"

            # Extract style attribute
            style_match = re.search(r'style="([^"]*)"', td_tag)
            subtitle_inline_style = style_match.group(1) if style_match else ""
        else:
            subtitle_inline_style = ""

        return {
            "title": title,
            "subtitle": subtitle,
            "title_class": title_class,
            "subtitle_class": subtitle_class,
            "title_style": title_inline_style,
            "subtitle_style": subtitle_inline_style,
            "style": style_content,
        }

    # Extract header data if specified
    if table_header_from == 1:
        header_data = extract_tab_header_and_style(gt1)
    elif table_header_from == 2:
        header_data = extract_tab_header_and_style(gt2)
    else:
        header_data = None

    # Build the header HTML if header data is available
    if header_data and (header_data["title"] or header_data["subtitle"]):
        header_html = f"""
        <style>
        {header_data["style"]}
        </style>
        """

        # Add title if it exists
        if header_data["title"]:
            header_html += f"""
            <div class="{header_data["title_class"]}"
                style="width:100%; {header_data["title_style"]}"
            >
                {header_data["title"]}
            </div>
            """

        # Add subtitle if it exists
        if header_data["subtitle"]:
            header_html += f"""
            <div class="{header_data["subtitle_class"]}"
                style="width:100%; {header_data["subtitle_style"]}">
                {header_data["subtitle"]}
            </div>
            """

        # Remove headers from the two tables
        gt1_no_header = gt1.tab_header(title=None, subtitle=None)  # type: ignore
        table_1_html = gt1_no_header.as_raw_html()
        gt2_no_header = gt2.tab_header(title=None, subtitle=None)  # type: ignore
        table_2_html = gt2_no_header.as_raw_html()

    else:
        header_html = ""
        table_1_html = gt1.as_raw_html()
        table_2_html = gt2.as_raw_html()

    double_table_html = f"""
    <div style="display: flex; justify-content: center; width: 100%;">
        <div id="mycombinedtable" style="display: inline-block; width: auto;">
            {header_html}
            <div style="overflow: auto; white-space: nowrap;">
                <div style="display: inline-block; margin-right: 1em;">
                    {table_1_html}
                </div>
                <div style="display: inline-block;">
                    {table_2_html}
                </div>
            </div>
        </div>
    </div>
    """

    double_table = GTCombinedLayout(double_table_html)

    if target:
        if target == "save":
            raise NotImplementedError(
                "At the moment, only notebook and browser display options are available."
            )
        else:
            double_table.show(target)

    return GTCombinedLayout(double_table_html)
