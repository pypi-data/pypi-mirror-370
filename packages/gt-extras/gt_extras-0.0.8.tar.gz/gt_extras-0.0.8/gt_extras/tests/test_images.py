from base64 import b64encode
from pathlib import Path

import pandas as pd
import polars as pl
import pytest
from conftest import assert_rendered_body
from great_tables import GT
from great_tables._text import Html

from gt_extras.images import FmtImage, add_text_img, gt_fmt_img_circle, img_header


def test_img_header_snapshot(snapshot):
    result = img_header(label="Test Label", img_url="https://example.com/image.png")
    assert snapshot == result


def test_img_header_basic():
    result = img_header(label="Test Label", img_url="https://example.com/image.png")

    assert isinstance(result, Html)
    assert "Test Label" in result.text
    assert "https://example.com/image.png" in result.text
    assert "height:60px;" in result.text
    assert "border-bottom:2px solid black;" in result.text
    assert "color:black;" in result.text


def test_img_header_custom_height_and_colors():
    result = img_header(
        label="Custom Label",
        img_url="https://example.com/custom.png",
        height=100,
        border_color="blue",
        text_color="red",
    )

    assert isinstance(result, Html)
    assert "Custom Label" in result.text
    assert "https://example.com/custom.png" in result.text
    assert "height:100px;" in result.text
    assert "border-bottom:2px solid blue;" in result.text
    assert "color:red;" in result.text


def test_img_header_custom_font_size():
    result = img_header(
        label="Font Size Test", img_url="https://example.com/font.png", font_size=20
    )

    assert isinstance(result, Html)
    assert "Font Size Test" in result.text
    assert "font-size:20px;" in result.text


def test_img_header_empty_label():
    result = img_header(label="", img_url="https://example.com/empty_label.png")

    assert isinstance(result, Html)
    assert "https://example.com/empty_label.png" in result.text
    assert "<div" in result.text
    assert "font-size:12px;" in result.text


def test_img_header_empty_url():
    result = img_header(label="Invalid URL Test", img_url="")

    assert isinstance(result, Html)
    assert "Invalid URL Test" in result.text
    assert 'src=""' in result.text


def test_img_header_no_border():
    result = img_header(
        label="No Border Test",
        img_url="https://example.com/no_border.png",
        border_color="transparent",
    )

    assert isinstance(result, Html)
    assert "No Border Test" in result.text
    assert "border-bottom:2px solid transparent;" in result.text


def test_add_text_img_snapshot(snapshot):
    result = add_text_img(
        text="Test Text",
        img_url="https://example.com/image.png",
        height=40,
        left=True,
    )
    assert snapshot == result


def test_add_text_img_basic():
    result = add_text_img(
        text="Test Text",
        img_url="https://example.com/image.png",
        height=40,
        left=False,
    )

    assert isinstance(result, str)
    assert "Test Text" in result
    assert "https://example.com/image.png" in result
    assert "height:40px;" in result
    assert "flex-direction:row-reverse;" in result


def test_add_text_img_left():
    result = add_text_img(
        text="Left Aligned Text",
        img_url="https://example.com/left_image.png",
        height=50,
        left=True,
    )

    assert isinstance(result, str)
    assert "Left Aligned Text" in result
    assert "https://example.com/left_image.png" in result
    assert "height:50px;" in result
    assert "flex-direction:row;" in result


def test_add_text_img_custom_gap():
    result = add_text_img(
        text="Custom Gap Text",
        img_url="https://example.com/custom_gap.png",
        height=30,
        gap=15.0,
        left=False,
    )

    assert isinstance(result, str)
    assert "Custom Gap Text" in result
    assert "https://example.com/custom_gap.png" in result
    assert "height:30px;" in result
    assert "gap:15.0px;" in result


def test_add_text_img_alt_text():
    result = add_text_img(
        text="Alt Text Test",
        img_url="https://example.com/image.png",
        height=40,
        left=True,
        alt_text="Example Alt Text",
    )

    assert isinstance(result, str)
    assert "Alt Text Test" in result
    assert "https://example.com/image.png" in result
    assert "alt='Example Alt Text'" in result
    assert "height:40px;" in result


def test_add_text_img_empty_text():
    result = add_text_img(
        text="",
        img_url="https://example.com/empty_text.png",
        height=30,
        left=True,
    )

    assert isinstance(result, str)
    assert "https://example.com/empty_text.png" in result
    assert "<div" in result
    assert "height:30px;" in result


def test_add_text_img_empty_url():
    result = add_text_img(
        text="Empty URL Test",
        img_url="",
        height=30,
        left=False,
    )

    assert isinstance(result, str)
    assert "Empty URL Test" in result
    assert "src=''" in result


def test_gt_fmt_img_circle_snapshot(snapshot):
    df = pd.DataFrame({"img": ["https://www.avatar1.png", "https://www.avatar2.png"]})
    gt_test = GT(df)

    res = gt_fmt_img_circle(gt_test, columns="img", encode=False, border_width="2px")
    assert_rendered_body(snapshot, gt=res)


def test_gt_fmt_img_circle_basic():
    df = pd.DataFrame({"img": ["test.png"]})
    gt_test = GT(df)

    result = gt_fmt_img_circle(gt_test, columns="img", encode=False, border_width="3px")
    html = result.as_raw_html()

    assert "border-radius: 50%;" in html
    assert "border-width: 3px;" in html
    assert "border-color: #0A0A0A;" in html
    assert "border-style: solid;" in html
    assert "test.png" in html


def test_gt_fmt_img_circle_custom_border():
    df = pd.DataFrame({"img": ["custom.jpg"]})
    gt_test = GT(df)

    result = gt_fmt_img_circle(
        gt_test,
        columns="img",
        encode=False,
        border_width=5,
        border_color="red",
        border_style="dashed",
    )
    html = result.as_raw_html()

    assert "border-width: 5px;" in html
    assert "border-color: red;" in html
    assert "border-style: dashed;" in html
    assert "border-radius: 50%;" in html


def test_gt_fmt_img_circle_custom_size():
    df = pd.DataFrame({"img": ["size_test.png"]})
    gt_test = GT(df)

    result = gt_fmt_img_circle(
        gt_test,
        columns="img",
        encode=False,
        height=80,
        width="80px",
        border_color="blue",
    )
    html = result.as_raw_html()

    assert "height: 80px;" in html
    assert "width: 80px;" in html
    assert "border-color: blue;" in html


def test_gt_fmt_img_circle_multiple_images():
    df = pd.DataFrame({"img": ["img1.png,img2.png"]})
    gt_test = GT(df)

    result = gt_fmt_img_circle(
        gt_test,
        columns="img",
        sep=" | ",
        encode=False,
        border_width="2px",
    )
    html = result.as_raw_html()

    assert "img1.png" in html
    assert "img2.png" in html
    assert " | " in html
    assert html.count("border-radius: 50%;") == 2


def test_gt_fmt_img_circle_no_border_defaults():
    """Test that without any border properties, no border styles are applied"""
    df = pd.DataFrame({"img": ["no_border.png"]})
    gt_test = GT(df)

    result = gt_fmt_img_circle(gt_test, columns="img", encode=False)
    html = result.as_raw_html()

    assert "border-radius: 50%;" in html
    assert "border-width:" not in html
    assert "border-color:" not in html


def test_gt_fmt_img_circle_partial_border_props():
    """Test that providing only border_color applies default width and style"""
    df = pd.DataFrame({"img": ["partial.png"]})
    gt_test = GT(df)

    result = gt_fmt_img_circle(
        gt_test, columns="img", encode=False, border_color="green"
    )
    html = result.as_raw_html()

    assert "border-color: green;" in html
    assert "border-width: 3px;" in html
    assert "border-style: solid;" in html


def test_gt_fmt_img_circle_custom_border_radius():
    df = pd.DataFrame({"img": ["radius_test.png"]})
    gt_test = GT(df)

    result = gt_fmt_img_circle(
        gt_test,
        columns="img",
        encode=False,
        border_radius="25%",
    )
    html = result.as_raw_html()

    assert "border-radius: 25%;" in html


def test_gt_fmt_img_circle_invalid_border_radius():
    df = pl.DataFrame({"img": ["test.png"], "height_col": ["80px"]})
    gt_test = GT(df)

    with pytest.raises(NotImplementedError):
        gt_fmt_img_circle(gt_test, columns="img", border_radius="4").as_raw_html()


def test_gt_fmt_img_circle_polars_expr_error():
    df = pl.DataFrame({"img": ["test.png"], "height_col": ["80px"]})
    gt_test = GT(df)

    with pytest.raises(NotImplementedError):
        gt_fmt_img_circle(
            gt_test,
            columns="img",
            height=pl.col("height_col"),
            border_width="2px",
        )


def test_gt_fmt_img_circle_as_latex():
    df = pl.DataFrame({"img": ["test.png"], "height_col": ["80px"]})
    gt_test = GT(df)

    with pytest.warns():
        gt_fmt_img_circle(gt_test, columns="img").as_latex()


def test_gt_fmt_img_circle_with_na_values():
    df = pd.DataFrame({"img": ["test.png", None, "another.jpg", pd.NA]})
    gt_test = GT(df)

    result = gt_fmt_img_circle(gt_test, columns="img", encode=False, border_width="2px")
    html = result.as_raw_html()

    assert "test.png" in html
    assert "another.jpg" in html
    assert isinstance(result, GT)


def strip_windows_drive(x):
    # this is a hacky approach to ensuring fmt_image path tests succeed
    # on our windows build. On linux root is just "/". On windows its a
    # drive name. Assumes our windows runner uses D:\
    return x.replace('src="D:\\', 'src="/')


@pytest.mark.parametrize(
    "ext,mime_type",
    [
        ("svg", "image/svg+xml"),
        ("jpg", "image/jpeg"),
        ("png", "image/png"),
    ],
)
def test_fmt_image_encode_param(tmpdir, ext, mime_type):
    content = "abc"
    p_img = Path(tmpdir) / f"some.{ext}"
    p_img.write_text(content)

    formatter = FmtImage(sep=" ", file_pattern=f"{{}}.{ext}", encode=True)
    res = formatter.to_html(f"{tmpdir}/some")

    b64_content = b64encode(content.encode()).decode()
    img_src = f"data:{mime_type};base64,{b64_content}"
    dst = formatter.SPAN_TEMPLATE.format(
        f'<img src="{img_src}" style="vertical-align: middle;">'
    )

    assert strip_windows_drive(res) == dst


@pytest.mark.parametrize(
    "url",
    ["http://posit.co/", "http://posit.co", "https://posit.co/", "https://posit.co"],
)
def test_fmt_image_path_http(url: str):
    formatter = FmtImage(encode=False, height=30, path=url, border_radius="50%")
    res = formatter.to_html("c")
    dst_img = '<img src="{}/c" style="height: 30px;border-radius: 50%;vertical-align: middle;">'.format(
        url.removesuffix("/")
    )
    dst = formatter.SPAN_TEMPLATE.format(dst_img)

    assert strip_windows_drive(res) == dst
