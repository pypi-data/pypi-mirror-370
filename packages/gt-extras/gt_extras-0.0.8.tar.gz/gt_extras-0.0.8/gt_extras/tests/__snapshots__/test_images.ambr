# serializer version: 1
# name: test_add_text_img_snapshot
  '''
  <div style='display:flex; flex-direction:row; align-items:center; gap:3.0px;'>
          <div style='flex-shrink: 0;'>
              <img src='https://example.com/image.png' alt=''
              style='height:40px; width:auto; object-fit:contain;'/>
          </div>
          <div style='flex-grow:1;'>
              Test Text
          </div>
      </div>
  '''
# ---
# name: test_gt_fmt_img_circle_snapshot
  '''
  <tbody class="gt_table_body">
    <tr>
      <td class="gt_row gt_left"><span style="white-space:nowrap;"><img src="https://www.avatar1.png" style="height: 2em;border-radius: 50%;border-width: 2px;border-color: #0A0A0A;border-style: solid;vertical-align: middle;"></span></td>
    </tr>
    <tr>
      <td class="gt_row gt_left"><span style="white-space:nowrap;"><img src="https://www.avatar2.png" style="height: 2em;border-radius: 50%;border-width: 2px;border-color: #0A0A0A;border-style: solid;vertical-align: middle;"></span></td>
    </tr>
  </tbody>
  '''
# ---
# name: test_img_header_snapshot
  Html(text='<div style="text-align:center;">\n        <img src="https://example.com/image.png" style="\n        height:60px;\n        object-fit:contain;\n        object-position: bottom;\n        border-bottom:2px solid black;"\n    />\n        <div style="\n        font-size:12px;\n        color:black;\n        text-align:center;\n        width:100%;\n    ">\n        Test Label\n    </div>\n    </div>')
# ---
